### PR TITLE
Fix 4089: Taxcloud with marketplace setup.

### LINKED ISSUE #3973: TaxCloud service not calculating tax
## Issue Description
When enabled, the TaxCloud service is not calculating tax in the cart. There are no errors on the client or server console and I am using the current API key

## Steps to Reproduce
1. Enable TaxCloud tax service
1. Add an item to cart and enter address
1. Observe that tax is not calculated

### Versions
```
Node: 8.9.4
NPM: 5.6.0
Meteor Node: 8.9.4
Meteor NPM: 5.6.0
Reaction CLI: 0.28.0
Reaction: 1.9.0
Reaction branch: release-1.9.0
Docker: 17.12.0-ce
```

### LINKED ISSUE #4089: Update Taxcloud tax method to work with Marketplace
Splitting this up into a separate issue from Avalara as these probably have separate solutions

This plugin needs to be "marketplace aware" in that it should calculate tax based on where the product is shipping from and where the marketplace is located.

Part of the scope of this ticket is investigating with TaxCloud to find out how they implement such a scenario. Possibly this might require that every marketplace shop signs up for its own TaxCloud account.

### PR DESCRIPTION
**Pull request by Akarshit**
Resolves #4089
Resolves #3973   
Impact: **minor**  
Type: **feature**

## Issue
1. Tax calculation in TaxCloud didn't support marketplace setup. It was using the address of the primary shop as origin.
2. Tax calculation was not working because, the tax calculation was happening in async, during which the default value of tax(0) was being used.
3. TaxCloud needs pincode in form of `{ Zip5: "12345", Zip4: "1234" }`, but we were sending it the pincode we had from the user.
4. Tax rate was not stored per shop(as this was not ready for marketplace.)
5. Taxcodes were not being fetched from TaxCloud's API. The job doing this was commented out. Also the tax-cloud API is now changed.

## Solution
1. A API call is made per shop, with all the items of the shop that were bought. So origin address changes with every shop's API call. Hence marketplace is supported
2. Tax calculation happens in sync.
3. This was not a major issue. To fix this used a regex to split the pincode and passed that to the API call.
4. Tax rate is stored per shop. But should we do this? Shouldn't it be per item?
5. Restored the job that fetches the taxcode and modified that code to reconcile the fetched taxcode with `TaxCode` schema.
6. Modified the code to reconcile with the new taxcloud API for tax-codes.

**Also I have removed the code which used to set `tax` to 0 by default. Because this was resetting the `tax` to 0, whenever cart was updated.** I have looked into the side-effects and couldn't spot any. But it would be better if someone can also think on this once.

## Breaking changes
None

### MarketPlace Setup
1. Login as admin(A), enable TaxCloud.
2. Invite a marketplace user(M).
3. Login as M and create a product, publish it.
4. Enable TaxCloud.
5. Set up your shop's address.
6. Add the product you created, and a product from the primary shop(Reaction Example Product) to the cart. 
7. Go to checkout screen. See the tax are being calculated correctly.
8. Buy the products.
9. Change either of the shop's address.
10. Again repeat the above steps.
11. See that the tax calculated is different.
12. Try checkout with different pincodes format(1234-12345, 12345, 123456789), see that all of them work.
13. Place the orders and see that tax is mentioned on the summary screen.
1. Enter wrong credentials for tax-cloud and save
1. Now add a product to the cart and checkout with it.
1. Go to the orders in the Action Bar. Notice it is marked with a red dot and is labelled `Tax not calculated`.

### Fetching of tax rates
1. Enable tax-cloud.
1. Set refresh period to `every 1 min`
2. Wait for 1 minute and after that the tax-code will be fetched.
3. To see this, go to PDP and edit product, select the title and go to Tax field. 
4. The dropdown for tax-codes should now be populated with the fetched tax codes.
5. Go back to tax-cloud settings, set the refresh period to `every 5 mins`, now tax-codes will be fetched every 5 mins.

### CODE REVIEW

**Comment by Akarshit** [comment]
@jshimko @zenweasel This works in a perfect world. Still working on how to force the user to do the pre-requisites for this to work like
1. Having same tax service enabled for primary and marketplace shop.
2. Having them have a shop address.

**Comment by zenweasel** [comment]
@Akarshit If you got this working when all things are set up properly let's go ahead and get this tested and merged and talk about how we do parent/child settings in a separate ticket. Probably want to get @spencern's input as well so we can create something that's reusable.

Shops are required to input an address before they can be made active so that part is already taken care of.

File: imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js [hunk 1]
```diff
@@ -1,8 +1,136 @@
...
+        cartItems: shopItemsMap[shopId],
+        origin,
+        destination,
+        cartID: cartToCalc._id,
+        deliveredBySeller: false
+      }
+    };
+
+    try {
+      const response = HTTP.post(url, request);
```
Comment (zenweasel) [inline imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js]: I thought we had talked about trying to see if we could make this call async?

**Comment by Akarshit** [comment]
Updated.

**Comment by zenweasel** [comment]
@Akarshit You have lint errors here

**Comment by zenweasel** [comment]
@Akarshit I keep getting this error on the server console even though I have the API configured:
```
02:43:23.816Z  WARN Reaction: TaxCloud API Key is required.
```

**Comment by zenweasel** [comment]
Getting this error on the server when checking out:
```
Exception in callback of async function: Error: Error calling taxcloud API
    at HTTP.post (imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js:138:17)
    at runWithEnvironment (packages/meteor.js:1240:24)
Exception in callback of async function: Error: Error calling taxcloud API
    at HTTP.post (imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js:138:17)
    at runWithEnvironment (packages/meteor.js:1240:24)
02:58:32.978Z ERROR Reaction: Error fetching tax rate from TaxCloud: An error has occurred proessing your request.  Please contact TaxCloud (code:2601)
02:58:33.025Z ERROR Reaction: Error fetching tax rate from TaxCloud: An error has occurred proessing your request.  Please contact TaxCloud (code:2601)
```

**Comment by zenweasel** [comment]
It doesn't look like taxes are showing up correctly in the Orders dashboard. I only see the tax for the second shop.

File: imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js [hunk 2]
```diff
@@ -1,8 +1,182 @@
...
+        cartID: cartToCalc._id,
+        deliveredBySeller: false
+      }
+    };
+    // pushing all the requests to array.
+    responsePromise.push(new Promise((resolve, reject) => {
+      HTTP.post(url, request, (error, response) => {
+        // ResponseType 3 is a successful call.
+        if (response.data.ResponseType !== 3) {
+          let errMsg = "Unable to access service. Check credentials.";
```
Comment (zenweasel) [inline imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js]: Are we sure that this is a credentials-based error?

File: imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js [hunk 3]
```diff
@@ -1,8 +1,182 @@
...
+}
+
+/**
+ * @name calculateTax
+ * @summary gets tax from the API and updates the cart.
+ * @method
+ * @param  {Object} pkgSettings tax-cloud settings
+ * @param  {Object} cartToCalc cart object from the database.
+ */
+function calculateTax(pkgSettings, cartToCalc) {
```
Comment (zenweasel) [inline imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js]: This function is over 100+ lines which makes it hard to follow/test. I wonder if we could try to break this down into more digestable chunks.

File: imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js (hunk 2 above)
Comment (Akarshit) [reply imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js]: Yes.

File: imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js (hunk 3 above)
Comment (Akarshit) [reply imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js]: Doing so.

**Comment by Akarshit** [comment]
>It doesn't look like taxes are showing up correctly in the Orders dashboard. I only see the tax for the second shop.

@zenweasel I think you should be only seeing the tax for the first shop. But the bigger question here is what should be ideal behavior. We did discuss on this before but couldn't come to a conclusion. Should the primary owner see both the taxes or should he see only the tax that corresponds to his shop?

**Comment by zenweasel** [comment]
@Akarshit Looks like you still have lint errors

**Comment by Akarshit** [comment]
Updated.

**Comment by Akarshit** [comment]
@zenweasel  Taxcloud seems to have changes their API for getting the TaxCodes. I am not we should be integrating that new API. There are 2 API's one for groups and one for tax-codes. I don't know if I should merge them or just get the tax-codes.

**Comment by Akarshit** [comment]
I found that there is a same version of the old API on a different URL. So will have to roll-back some changes that I have made.

**Comment by Akarshit** [comment]
@zenweasel Updated.

**Comment by spencern** [comment]
@zenweasel is this a PR that we want to merge into 1.14.0 seeing that we're also planning to move the `tax-cloud` plugin out of core for that release?

**Comment by zenweasel** [comment]
@spencern Yes, since the work is already done on this, I was hoping to get this merged before migrating the code out of core rather than PR into the new repo.

**Comment by spencern** [comment]
@zenweasel is the goal still to remove the TaxCloud package for 1.14.0?

**Comment by zenweasel** [comment]
@spencern Yes. As soon as this PR is merged I or @Akarshit can submit a PR to remove TaxCloud.

**Comment by zenweasel** [comment]
@Akarshit When I test this with the primary shop in CA and the Marketplace shop in Portland Oregon (which does not have a sales tax) I still see a tax being charged.

Also not seeing taxes reflected in the detail view.

**Comment by Akarshit** [comment]
I missed updating the product detail view, will modify it.

**Comment by Akarshit** [comment]
@zenweasel I looked more into why the product in Oregon was being taxed, and I found that it depends on the tax-code. So for the taxcode 0000(our default) the tax is 1.02/item and for the taxcode 11011 the tax is 0.
I don't think we can do much about this.

**Comment by Akarshit** [comment]
@zenweasel Also about the tax not being shown on the "detail view", I think that is by design since we don't show discounts and shipping also.
Also the "subtotal" internally in the code refers to raw price of item.quantity * item.cost.

But an argument could be made that shipping and discount are not present per item(yet) and taxes are.

**Comment by zenweasel** [comment]
That's fundamentally not how sales taxes work

**Comment by zenweasel** [comment]
Also, taxes are absolutely supposed to show in detail view

**Comment by zenweasel** [comment]
Try doing a purchase where both the shop and the shipping address are in OR. If that is still being charged tax we've got a serious problem

**Comment by Akarshit** [comment]
@zenweasel 
OK, I will add tax amounts to the detail view. 

Intra Oregon, taxes are 0 irrespective of TIC.

**Comment by zenweasel** [comment]
@Akarshit re: Oregon. 
Ok, that makes sense. I guess it charges based on shipping address + shipper. So it's working how I would expect. So disregard my previous note.

**Comment by Akarshit** [comment]
@zenweasel I have updated the PR with the requested changes.

**Comment by zenweasel** [comment]
@Akarshit You've got two lint errors here

**Comment by zenweasel** [comment]
I'm not seeing tax codes ever updated/populated. It was running all last night

**Comment by Akarshit** [comment]
@zenweasel Yes that would be the case because by default the tax codes will be downloaded "every 7 days".

As you suggested I will add a field to the form to change this "refresh period".

**Comment by Akarshit** [comment]
@zenweasel I have added a field to set the `refreshInterval` to the tax-cloud settings form.

File: imports/plugins/core/taxes/server/methods/methods.js [hunk 4]
```diff
@@ -108,7 +108,7 @@ export const methods = {
     check(cartId, String);
     check(options, {
       taxRatesByShop: Match.OneOf(undefined, null, Object),
-      itemsWithTax: [Object],
+      itemsWithTax: [Match.OneOf(Object, undefined)],
```
Comment (spencern) [inline imports/plugins/core/taxes/server/methods/methods.js]: fyi you can do
`Match.Optional(Pattern)` for `Match.OneOf(Pattern, undefined)`
and
`Match.Maybe(Pattern)` for `Match.OneOf(Pattern, undefined, null)`

**Comment by zenweasel** [comment]
@Akarshit I found the one issue with address validation but I think that's because the scope of this PR (and probably dependent on the Shippo PR to be merged first)

**Comment by spencern** [comment]
@zenweasel did you create a ticket for the address validation issue you found here?

**Comment by zenweasel** [comment]
@spencern No, I will create one on the new Taxcloud repo once that's created.

**Comment by zenweasel** [comment]
@spencern If possible I would like to merge this today/soon so we can work on moving it out of core.

**Comment by stunjiturner** [comment]
Can I checkout the 1.4 branch, and test it for tax calculations, for marketplace and single store.

**Comment by Akarshit** [comment]
@stunjiturner Exactly right now? then yes.
But there is a PR to remove the tax-cloud into a custom plugin. Once that gets merged, you will have to install the tax-cloud as according to https://github.com/reaction-contrib/meteor-taxes-taxcloud

**Comment by stunjiturner** [comment]
I am not able to edit the tax information for an individual product, the tax shows in the cart review, but it seems incorrect because when I use a PA address, there should not be a tax on clothes, but since I cannot add a tax code to a product, in the admin section the calculations can't ever be right.

screen shot 2018-07-16 at 3 35 56 am

**Comment by zenweasel** [comment]
@stunjiturner this has already been merged. You should open a new issue on the contrib repo

**Comment by stunjiturner** [comment]
can I use the new plugin with the master repo version 1.3.1?

**Comment by Akarshit** [comment]
@stunjiturner No.

**Comment by stunjiturner** [comment]
its not syncing the tax codes, nor are the toggles working for me, going to try fire fox

**Comment by stunjiturner** [comment]
yes it works, now
